File: Data Structures & Algorithms/palindromic-substrings/submission-2.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        res = 0
        p1,p2 = 0,0
        while p1<len(s) and p2<len(s):
            tp1,tp2 = p1,p2
            while tp1>=0 and tp2<len(s) and s[tp1] == s[tp2]:
                res+=1
                tp1-=1
                tp2+=1
            if p1 == p2:
                p2+=1
            else:
                p1+=1
        return res

File: Data Structures & Algorithms/longest-happy-string/submission-2.cpp
class Solution {
    struct fpair{
        int freq;
        char c;
    };

    struct fpairComp{
        bool operator()(const fpair& a, const fpair& b){
            return a.freq < b.freq;
        }
    };
public:
    string longestDiverseString(int a, int b, int c) {

        std::priority_queue<fpair,vector<fpair>,fpairComp> pq(fpairComp{});
        if (a > 0){pq.push(fpair{a,'a'});}
        if (b > 0){pq.push(fpair{b,'b'});}
        if (c > 0){pq.push(fpair{c,'c'});}

        std::string res = "";
        while (pq.size()>0){
            auto [freq,c] = pq.top();
            pq.pop();
            int l = res.size();
            if (l >= 2 && res[l-2] == res[l-1] && res[l-1] == c){
                if (pq.size() > 0){
                    auto [freq2,c2] = pq.top();
                    pq.pop();
                    res += c2;
                    if (freq2 > 1){
                        pq.push(fpair{freq2-1,c2});
                    }
                } else {
                    return res;
                }
            } else{
                res += c;
                freq -= 1;
            }
            if (freq > 0){pq.push(fpair{freq, c});}
        }
        return res;
        
    }
};